File: Kalorie/Kalorie/Core/Auth/PendingMergeSnapshotStore.swift
//
//  PendingMergeSnapshotStore.swift
//  Kalorie
//
//  Created by Josef Antoni on 07.08.2026.
//

import Foundation

struct PendingMergeSnapshot: Codable {

    // MARK: - Properties

    let sourceAnonymousUserId: String
    let foodConsumed: [FoodConsumedDTO]
    let favouriteFoods: [FavouriteFoodDTO]

    // MARK: - Init

    init(sourceAnonymousUserId: String, foodConsumed: [FoodConsumedDTO], favouriteFoods: [FavouriteFoodDTO]) {
        self.sourceAnonymousUserId = sourceAnonymousUserId
        self.foodConsumed = foodConsumed
        self.favouriteFoods = favouriteFoods
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        sourceAnonymousUserId = try container.decode(String.self, forKey: .sourceAnonymousUserId)
        foodConsumed = try container.decode([FoodConsumedDTO].self, forKey: .foodConsumed)
        favouriteFoods = try container.decodeIfPresent([FavouriteFoodDTO].self, forKey: .favouriteFoods) ?? []
    }
}

protocol PendingMergeSnapshotStoreProtocol {
    func save(_ snapshot: PendingMergeSnapshot) throws
    func load() throws -> PendingMergeSnapshot?
    func delete() throws
}

struct PendingMergeSnapshotStore: PendingMergeSnapshotStoreProtocol {

    // MARK: - Properties

    private var fileURL: URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("pending-merge.json")
    }

    // MARK: - Functions

    func save(_ snapshot: PendingMergeSnapshot) throws {
        try JSONEncoder().encode(snapshot).write(to: fileURL, options: .atomic)
    }

    func load() throws -> PendingMergeSnapshot? {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return nil }
        return try JSONDecoder().decode(PendingMergeSnapshot.self, from: Data(contentsOf: fileURL))
    }

    func delete() throws {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return }
        try FileManager.default.removeItem(at: fileURL)
    }
}

#if DEBUG
final class PendingMergeSnapshotStoreFake: PendingMergeSnapshotStoreProtocol {

    // MARK: - Properties

    var stubbedSnapshot: PendingMergeSnapshot?
    var saveError: Error?
    private(set) var deleteCallCount = 0

    // MARK: - Functions

    func save(_ snapshot: PendingMergeSnapshot) throws {
        if let saveError { throw saveError }
        stubbedSnapshot = snapshot
    }

    func load() throws -> PendingMergeSnapshot? {
        stubbedSnapshot
    }

    func delete() throws {
        deleteCallCount += 1
        stubbedSnapshot = nil
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/auth/PendingMergeSnapshotStore.kt
package antoni.kalorie.core.auth

import antoni.kalorie.core.networking.FavouriteFoodDTO
import antoni.kalorie.core.networking.FoodConsumedDTO
import antoni.kalorie.core.networking.FoodItemPersonalPortionsDTO
import antoni.kalorie.core.networking.MyCreatedMealDTO
import java.io.File
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class PendingMergeSnapshot(
    val sourceAnonymousUserId: String,
    val foodConsumed: List<FoodConsumedDTO>,
    val favouriteFoods: List<FavouriteFoodDTO> = emptyList(),
    val myCreatedMeals: List<MyCreatedMealDTO> = emptyList(),
    val foodItemPortions: List<FoodItemPersonalPortionsDTO> = emptyList(),
)

interface PendingMergeSnapshotStoreProtocol {
    fun save(snapshot: PendingMergeSnapshot)
    fun load(): PendingMergeSnapshot?
    fun delete()
}

class PendingMergeSnapshotStore(directory: File) : PendingMergeSnapshotStoreProtocol {

    // MARK: - Properties

    private val file = File(directory, "pending-merge.json")
    private val json = Json { ignoreUnknownKeys = true }

    // MARK: - Functions

    override fun save(snapshot: PendingMergeSnapshot) {
        val temporaryFile = File(file.parentFile, "pending-merge.json.tmp")
        temporaryFile.writeText(json.encodeToString(PendingMergeSnapshot.serializer(), snapshot))
        if (!temporaryFile.renameTo(file)) {
            file.delete()
            check(temporaryFile.renameTo(file)) { "Could not replace the pending merge snapshot" }
        }
    }

    override fun load(): PendingMergeSnapshot? {
        if (!file.exists()) return null
        return json.decodeFromString(PendingMergeSnapshot.serializer(), file.readText())
    }

    override fun delete() {
        if (!file.exists()) return
        check(file.delete()) { "Could not delete the pending merge snapshot" }
    }
}
